src/aead/gcm/ffi.rs

```diff
@@ -0,0 +1,186 @@
+
+#[cfg(any(
+    target_arch = "aarch64",
+    target_arch = "arm",
+    target_arch = "x86",
+    target_arch = "x86_64"
+))]
+macro_rules! htable_new {
+    ( $name:ident, $input:expr, $cpu_features:expr ) => {{
+        use crate::aead::gcm::ffi::HTable;
+        prefixed_extern! {
+            fn $name(HTable: &mut HTable, h: &[u64; 2]);
+        }
+        HTable::new($name, $input)
+    }};
+}
+
+#[cfg(any(
+    target_arch = "aarch64",
+    target_arch = "arm",
+    target_arch = "x86",
+    target_arch = "x86_64"
+))]
+macro_rules! gmult {
+    ( $name:ident, $xi:expr, $h_table:expr, $cpu_features:expr ) => {{
+        use crate::aead::gcm::ffi::{HTable, Xi};
+        prefixed_extern! {
+            fn $name(xi: &mut Xi, Htable: &HTable);
+        }
+        $h_table.gmult($name, $xi, $cpu_features)
+    }};
+}
+
+/// SAFETY:
+///  * The function `$name` must meet the contract of the `f` paramweter of
+///    `ghash()`.
+#[cfg(any(
+    target_arch = "aarch64",
+    target_arch = "arm",
+    target_arch = "x86",
+    target_arch = "x86_64"
+))]
+macro_rules! ghash {
+    ( $name:ident, $xi:expr, $h_table:expr, $input:expr, $cpu_features:expr ) => {{
+        use crate::aead::gcm::ffi::{HTable, Xi};
+        prefixed_extern! {
+            fn $name(
+                xi: &mut Xi,
+                Htable: &HTable,
+                inp: *const u8,
+                len: crate::c::NonZero_size_t,
+            );
+        }
+        $h_table.ghash($name, $xi, $input, $cpu_features)
+    }};
+}
+
+/// SAFETY:
+///   * `f` must read `len` bytes from `inp`; it may assume
+///     that `len` is a (non-zero) multiple of `BLOCK_LEN`.
+///   * `f` may inspect CPU features.
+#[cfg(any(
+    target_arch = "aarch64",
+    target_arch = "arm",
+    target_arch = "x86",
+    target_arch = "x86_64"
+))]
+impl HTable {
+    pub(super) unsafe fn new(
+        init: unsafe extern "C" fn(HTable: &mut HTable, h: &[u64; 2]),
+        value: &[u64; 2],
+    ) -> Self {
+        let mut r = Self {
+            Htable: [U128 { hi: 0, lo: 0 }; HTABLE_LEN],
+        };
+        unsafe { init(&mut r, value) };
+        r
+    }
+
+    pub(super) unsafe fn gmult(
+        &self,
+        f: unsafe extern "C" fn(xi: &mut Xi, h_table: &HTable),
+        xi: &mut Xi,
+        _cpu_features: crate::cpu::Features,
+    ) {
+        unsafe { f(xi, self) }
+    }
+
+    pub(super) unsafe fn ghash(
+        &self,
+        f: unsafe extern "C" fn(
+            xi: &mut Xi,
+            Htable: &HTable,
+            inp: *const u8,
+            len: crate::c::NonZero_size_t,
+        ),
+        xi: &mut Xi,
+        input: &[[u8; BLOCK_LEN]],
+        cpu_features: crate::cpu::Features,
+    ) {
+        use crate::polyfill::slice;
+        use core::num::NonZeroUsize;
+
+        let input = slice::flatten(input);
+
+        let input_len = match NonZeroUsize::new(input.len()) {
+            Some(len) => len,
+            None => {
+                return;
```

Comment:
This is currently unreachable (before and after this commit), but it might not always be unreachable, so we'll leave it as-is. In the future we should write a test.